File: Leetcode100/leetcode100_python/MedianofTwoSortedArrays.py
from typing import List

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        m = len(nums1)
        n = len(nums2)
        if m > n:
            nums1, nums2 = nums2, nums1
            m, n = n, m
        
        begin = 0
        end = m
        while begin <= end:
            i = (begin + end) // 2
            j = (m + n + 1) // 2 - i
            if i > begin and nums1[i - 1] > nums2[j]:
                end = i - 1
            elif i < end and nums2[j - 1] > nums1[i]:
                begin = i + 1
            else:
                max_left = 0
                if i == 0:
                    max_left = nums2[j - 1]
                elif j == 0:
                    max_left = nums1[i - 1]
                else:
                    max_left = max(nums1[i - 1], nums2[j - 1])
                if (m + n) % 2 == 1:
                    return max_left
                
                min_right = 0
                if i == m:
                    min_right = nums2[j]
                elif j == n:
                    min_right = nums1[i]
                else:
                    min_right = min(nums1[i], nums2[j])
                return (max_left + min_right) / 2

        return 0

if __name__ == "__main__":
    nums1 = [int(num) for num in input().split()]
    nums2 = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.findMedianSortedArrays(nums1, nums2))

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/median_of_two_sorted_arrays/MedianofTwoSortedArrays.java
package main.java.com.jiang.leetcode100.median_of_two_sorted_arrays;

import java.util.Scanner;

/**
 * 复杂度不符合要求
 */
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {

        int n1 = nums1.length, n2 = nums2.length;
        int mid = (n1 + n2) / 2;
        
        int n = 0;
        int i = 0, j = 0;
        int pre = 0, num = 0;
        while(n <= mid) {
            pre = num;
            if(j >= n2 || (i < n1 && nums1[i] <= nums2[j])) {
                num = nums1[i];
                i++;
            } else {
                num = nums2[j];
                j++;
            }
            n++;
        }

        double result = 0;
        if((n1 + n2) % 2 == 0) result = (pre + num) / 2.0;
        else result = num;

        return result;
    }
}

public class MedianofTwoSortedArrays {
    public static void main(String[] args) {

        Scanner scan = new Scanner(System.in);
        String str1 = scan.nextLine();
        String str2 = scan.nextLine();
        scan.close();

        String[] snums1 = str1.split(" ");
        String[] snums2 = str2.split(" ");

        int[] nums1 = new int[snums1.length];
        int[] nums2 = new int[snums2.length];

        for(int i = 0; i < snums1.length; i++) {
            nums1[i] = Integer.parseInt(snums1[i]);
        }

        for(int i = 0; i < snums2.length; i++) {
            nums2[i] = Integer.parseInt(snums2[i]);
        }

        System.out.println(new Solution().findMedianSortedArrays(nums1, nums2));

    }
}